capsules/src/net/udp/driver.rs

```diff
@@ -0,0 +1,473 @@
+                app.rx_callback = callback;
+                ReturnCode::SUCCESS
+            }),
+            1 => self.do_with_app(app_id, |app| {
+                app.tx_callback = callback;
+                ReturnCode::SUCCESS
+            }),
+            _ => ReturnCode::ENOSUPPORT,
+        }
+    }
+
+    /// UDP control
+    ///
+    /// ### `command_num`
+    ///
+    /// - `0`: Driver check.
+    /// - `1`: Get the interface list
+    ///        app_cfg (out): 16 * `n` bytes: the list of interface IPv6 addresses, length
+    ///                       limited by `app_cfg` length.
+    /// - `2`: Transmit payload
+    fn command(&self, command_num: usize, arg1: usize, _: usize, appid: AppId) -> ReturnCode {
+        match command_num {
+            0 => ReturnCode::SUCCESS,
+
+            //  Writes the requested number of network interface addresses
+            // `arg1`: number of interfaces requested that will fit into the buffer
+            1 => self.do_with_cfg_mut(appid, arg1 * mem::size_of::<IPAddr>(), |cfg| {
+                let n_ifaces_to_copy = cmp::min(arg1, INTERFACES.len());
+                let iface_size = mem::size_of::<IPAddr>();
+                for i in 0..n_ifaces_to_copy {
+                    cfg[i * iface_size .. (i+1) * iface_size].copy_from_slice(&INTERFACES[i].0);
+                }
+                // Returns total number of interfaces
+                ReturnCode::SuccessWithValue { value: INTERFACES.len() }
+            }),
+
+            // Transmits UDP packet stored in
+            2 => {
+                self.do_with_app(appid, |app| {
+                    if app.pending_tx.is_some() {
+                        // Cannot support more than one pending tx per process.
+                        return ReturnCode::EBUSY;
+                    }
+                    let next_tx = app.app_cfg.as_ref().and_then(|cfg| {
+                        if cfg.len() != 2 * mem::size_of::<IPAddrPort>() {
+                            // debug!("cfg len is {:?}, needed at least {:?}", cfg.len(), 2 * mem::size_of::<IPAddrPort>());
+                            return None;
+                        }
+
+                        debug!("{:?}", self.parse_ip_port_pair(&cfg.as_ref()[mem::size_of::<IPAddrPort>()..]));
+
+                        if let (Some(dst), Some(src)) = (self.parse_ip_port_pair(&cfg.as_ref()[mem::size_of::<IPAddrPort>()..]),
+                                                         self.parse_ip_port_pair(&cfg.as_ref()[..mem::size_of::<IPAddrPort>()])) {
+                            Some([src, dst])
+                        } else {
+                            None
+                        }
+                    });
+                    if next_tx.is_none() {
+                        debug!("TX IS NONE");
+                        return ReturnCode::EINVAL;
+                    }
+                    app.pending_tx = next_tx;
+
+                    self.do_next_tx_sync(appid)
```

Comment:
I'm not sure how an app would use this interface. `SUCCESS` can either mean "the tx was queued" or "the packet was sent to the radio", and only the former gets a callback when the packet is sent to the radio, correct? How does the app know whether to wait for the callback?

Reply:
I believe a callback is triggered in either case. Regardless of whether the tx was queued or sent immediately, the send_done() callback passed up the radio stack in the kernel is always passed up to userspace via app.tx_callback on line 434ish. @ptcrews is this correct?  

The current implementation (which can be seen in the PR to tock/libtock-c) always waits for this callback.

Reply:
Ok the app is going to wait for the `send_done` callback, and it might get a `packet sent to radio` callback as well, which it likely can just ignore.

I do think it makes sense to add a return value to `SUCCESS` to indicate if the packet was sent to the radio immediately, in case the app wants to use that opportunity to send another packet immediately (rather than having to wait for send done).

Reply:
The app needs to wait for the `packet sent to radio` callback to make sure there was not an error,
if it ignores that callback completely there is a possibility the `send_done` callback never comes, and the app does not know why. I agree that adding a return value to indicate if the packet was immediately sent to the radio is a good idea, implementing that now.

Reply:
Yeah, there's no such thing as the packet being sent immediately: starting a transmission will require a bunch of asynchronous SPI operations and a failure could occur at any time. Tracking individual packets when there are multiple in flight starts to get tricky, especially if you can't maintain references to them. 

The major reason you'd want to queue as you describe is performance: with an explicit callback, there's a period when the radio is idle, between completion and the callback being issued. For any modern LAN technology this is a real concern, but for low-power link layers, with how fast MCUs, it isn't. But in cases when you do want to be able to enqueue multiple packets to a lower layer you probably want a different trait and intermediate queues.

Reply:
@phil-levis So you are saying that it probably makes more sense for userland apps to always wait for the `send_done` callback anyway before queuing another packet anyway? I buy that, and that means we would not need to change how udp.c is currently implemented in libtock-c. 

However, it still seems that it might be valuable to tell the userland app whether to expect a `packet sent to radio` callback, which could always contain Currently, it seems that there is no way for an app to determine if a packet was sent synchronously or asynchronously, and thus no way for an app to know whether the `packet sent to radio` callback might occur.

Reply:
So how would you recommend we synchronously return an error in the case that the packet is copied into a radio buffer in CPU memory, then sent later once currently in-progress transmissions by other apps finish? In this case the only way I can envision returning these errors would be via a callback, but in the instance that a packet is sent immediately because no other transmissions are in progress it seems much easier to simply directly return any errors that may occur, rather than setting up a callback to pass those errors.

At the end of the day, there are errors which can be generated in the process of passing a packet from the udp driver down to the radio, and if that occurs, we need to inform the user that happened. I agree that the caller should not assume a packet has been sent until it receives a `send_done` callback -- the only purpose of the `packet sent to radio callback` is really to pass back up any errors that might occur before the packet reaches the radio, so that the calling app is not left ignorant to what errors occurred.

Agreed that the different return values of do_next_tx_sync() need to be better documented, I will push some changes with additional documentation within the hour.

Reply:
I mean, all packets are sent later. Among other things, you usually have to put the radio into a transmit state. I'll take a look at the exact cases for synchronous and asynchronous before commenting more later tonight.

Reply:
Ah, OK. It looks like the terminology in the comments is a little confusing. The distinction is not about synchronous vs. asynchronous transmission, but rather whether the function signals a callback. The async functions simply call the sync ones, and based on the return value of the sync ones, decide whether to issue a callback (on an error).

More generally, yeah, some errors can be detected synchronously and should be returned immediately. Others might only be discovered asynchronously and so returned in the callback.

Reply:
I agree it would be easier to read the code if those were named differently (as they both use asynchronous operations). Maybe using "immediate" and "queued"?